### PR TITLE
cmap: return the oldest records from each shards first on Keys()

### PR DESCRIPTION
**Pull request by AdoAdoAdo**

### CODE REVIEW

File: concurrent_map.go [hunk 1]
```diff
@@ -109,7 +127,7 @@ func (m *ConcurrentMap) SetIfAbsent(key string, value interface{}) bool {
 	shard.Lock()
 	_, ok := shard.items[key]
 	if !ok {
-		shard.items[key] = value
+		shard.items[key] = &valWithIndex{val: value, arrayIdx: shard.idxAdd}
```
Comment (raduchis) [inline concurrent_map.go:136]: !ok means there is no item for the key?

Comment (AdoAdoAdo) [reply concurrent_map.go:136]: right, this method needs to add the item only if it is missing

Comment (raduchis) [reply concurrent_map.go:136]: ok

**Review by raduchis — APPROVED** [review]

File: concurrent_map.go [hunk 2]
```diff
@@ -64,7 +69,10 @@ func (m *ConcurrentMap) Set(key string, value interface{}) {
 	shard := m.GetShard(key)
 	shard.Lock()
 	_, ok := shard.items[key]
-	shard.items[key] = value
+	shard.items[key] = &valWithIndex{
```
Comment (raduchis) [inline concurrent_map.go:72]: If the value is found, it should be removed first and then readded on the latest position.

**Review by raduchis — CHANGES_REQUESTED** [review]

File: concurrent_map.go (hunk 2 above)
Comment (AdoAdoAdo) [reply concurrent_map.go:72]: done

**Review by raduchis — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**